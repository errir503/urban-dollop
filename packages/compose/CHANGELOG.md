<!-- Learn how to maintain this file at https://github.com/WordPress/gutenberg/tree/HEAD/packages#maintaining-changelogs. -->

## Unreleased

## 5.0.0 (2021-07-29)

### Breaking Change

-   Upgraded React components to work with v17.0 ([#29118](https://github.com/WordPress/gutenberg/pull/29118)). There are no new features in React v17.0 as explained in the [blog post](https://reactjs.org/blog/2020/10/20/react-v17.html).

## 4.2.0 (2021-07-21)

### Deprecations

-   `withState` HOC has been deprecated. Use `useState` hook instead.

### New Features

-   Publish TypeScript types.

## 4.1.0 (2021-05-20)

## 4.0.0 (2021-05-14)

### Breaking Changes

-   Drop support for Internet Explorer 11 ([#31110](https://github.com/WordPress/gutenberg/pull/31110)). Learn more at https://make.wordpress.org/core/2021/04/22/ie-11-support-phase-out-plan/.
-   Increase the minimum Node.js version to v12 matching Long Term Support releases ([#31270](https://github.com/WordPress/gutenberg/pull/31270)). Learn more at https://nodejs.org/en/about/releases/.

## 3.25.0 (2021-03-17)

## 3.24.0 (2021-01-21)

### New Features

-   Add the `useIsomorphicLayoutEffect` hook.

## 3.4.0 (2019-06-12)

### New Features

-   Add the `useMediaQuery` and `useReducedMotion` hooks.

## 3.0.0 (2018-11-15)

### Breaking Changes

-   `remountOnPropChange` has been removed.

## 2.1.2 (2018-11-09)

## 2.1.1 (2018-11-09)

## 2.1.0 (2018-10-29)

### Deprecation

-   `remountOnPropChange` has been deprecated.

## 2.0.5 (2018-10-19)

## 2.0.4 (2018-10-18)

## 2.0.0 (2018-09-05)

### Breaking Change

-   Change how required built-ins are polyfilled with Babel 7 ([#9171](https://github.com/WordPress/gutenberg/pull/9171)). If you're using an environment that has limited or no support for ES2015+ such as lower versions of IE then using [core-js](https://github.com/zloirock/core-js) or [@babel/polyfill](https://babeljs.io/docs/en/next/babel-polyfill) will add support for these methods.
