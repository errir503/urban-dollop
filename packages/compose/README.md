# Compose

The `compose` package is a collection of handy [Hooks](https://reactjs.org/docs/hooks-intro.html) and [Higher Order Components](https://facebook.github.io/react/docs/higher-order-components.html) (HOCs) you can use to wrap your WordPress components and provide some basic features like: state, instance id, pure...

The `compose` function is an alias to [flowRight](https://lodash.com/docs/#flowRight) from Lodash. It comes from functional programming, and allows you to compose any number of functions. You might also think of this as layering functions; `compose` will execute the last function first, then sequentially move back through the previous functions passing the result of each function upward.

An example that illustrates it for two functions:

```js
const compose = ( f, g ) => x
    => f( g( x ) );
```

Here's a simplified example of **compose** in use from Gutenberg's [`PluginSidebar` component](https://github.com/WordPress/gutenberg/blob/HEAD/packages/edit-post/src/components/sidebar/plugin-sidebar/index.js):

Using compose:

```js
const applyWithSelect = withSelect( ( select, ownProps ) => {
	return doSomething( select, ownProps );
} );
const applyWithDispatch = withDispatch( ( dispatch, ownProps ) => {
	return doSomethingElse( dispatch, ownProps );
} );

export default compose(
	withPluginContext,
	applyWithSelect,
	applyWithDispatch
)( PluginSidebarMoreMenuItem );
```

Without `compose`, the code would look like this:

```js
const applyWithSelect = withSelect( ( select, ownProps ) => {
	return doSomething( select, ownProps );
} );
const applyWithDispatch = withDispatch( ( dispatch, ownProps ) => {
	return doSomethingElse( dispatch, ownProps );
} );

export default withPluginContext(
	applyWithSelect( applyWithDispatch( PluginSidebarMoreMenuItem ) )
);
```

## Installation

Install the module

```bash
npm install @wordpress/compose --save
```

_This package assumes that your code will run in an **ES2015+** environment. If you're using an environment that has limited or no support for such language features and APIs, you should include [the polyfill shipped in `@wordpress/babel-preset-default`](https://github.com/WordPress/gutenberg/tree/HEAD/packages/babel-preset-default#polyfill) in your code._

## API

For more details, you can refer to each Higher Order Component's README file. [Available components are located here.](https://github.com/WordPress/gutenberg/tree/HEAD/packages/compose/src)

<!-- START TOKEN(Autogenerated API docs) -->

### compose

Composes multiple higher-order components into a single higher-order component. Performs right-to-left function
composition, where each successive invocation is supplied the return value of the previous.

This is just a re-export of `lodash`'s `flowRight` function.

_Related_

-   <https://docs-lodash.com/v4/flow-right/>

### createHigherOrderComponent

Given a function mapping a component to an enhanced component and modifier
name, returns the enhanced component augmented with a generated displayName.

_Parameters_

-   _mapComponentToEnhancedComponent_ `HigherOrderComponent< TInnerProps, TOuterProps >`: Function mapping component to enhanced component.
-   _modifierName_ `string`: Seed name from which to generated display name.

_Returns_

-   `HigherOrderComponent< TInnerProps, TOuterProps >`: Component class with generated display name assigned.

### ifCondition

Higher-order component creator, creating a new component which renders if
the given condition is satisfied or with the given optional prop name.

_Usage_

```ts
type Props = { foo: string };
const Component = ( props: Props ) => <div>{ props.foo }</div>;
const ConditionalComponent = ifCondition(
	( props: Props ) => props.foo.length !== 0
)( Component );
<ConditionalComponent foo="" />; // => null
<ConditionalComponent foo="bar" />; // => <div>bar</div>;
```

_Parameters_

-   _predicate_ `( props: TProps ) => boolean`: Function to test condition.

_Returns_

-   `HigherOrderComponent< TProps, TProps >`: Higher-order component.

### pure

Given a component returns the enhanced component augmented with a component
only rerendering when its props/state change

_Type_

-   `SimpleHigherOrderComponent`

### useAsyncList

React hook returns an array which items get asynchronously appended from a source array.
This behavior is useful if we want to render a list of items asynchronously for performance reasons.

_Parameters_

-   _list_ `T[]`: Source array.
-   _config_ `AsyncListConfig`: Configuration object.

_Returns_

-   `T[]`: Async array.

### useConstrainedTabbing

In Dialogs/modals, the tabbing must be constrained to the content of
the wrapper element. This hook adds the behavior to the returned ref.

_Usage_

```js
import { useConstrainedTabbing } from '@wordpress/compose';

const ConstrainedTabbingExample = () => {
	const constrainedTabbingRef = useConstrainedTabbing();
	return (
		<div ref={ constrainedTabbingRef }>
			<Button />
			<Button />
		</div>
	);
};
```

_Returns_

-   `import('react').RefCallback<Element>`: Element Ref.

### useCopyOnClick

> **Deprecated**

Copies the text to the clipboard when the element is clicked.

_Parameters_

-   _ref_ `import('react').RefObject<string | Element | NodeListOf<Element>>`: Reference with the element.
-   _text_ `string|Function`: The text to copy.
-   _timeout_ `[number]`: Optional timeout to reset the returned state. 4 seconds by default.

_Returns_

-   `boolean`: Whether or not the text has been copied. Resets after the timeout.

### useCopyToClipboard

Copies the given text to the clipboard when the element is clicked.

_Parameters_

-   _text_ `string | (() => string)`: The text to copy. Use a function if not already available and expensive to compute.
-   _onSuccess_ `Function`: Called when to text is copied.

_Returns_

-   `import('react').Ref<TElementType>`: A ref to assign to the target element.

### useDebounce

Debounces a function with Lodash's `debounce`. A new debounced function will
be returned and any scheduled calls cancelled if any of the arguments change,
including the function to debounce, so please wrap functions created on
render in components in `useCallback`.

_Related_

-   <https://docs-lodash.com/v4/debounce/>

_Parameters_

-   _fn_ `TFunc`: The function to debounce.
-   _wait_ `[number]`: The number of milliseconds to delay.
-   _options_ `[import('lodash').DebounceSettings]`: The options object.

_Returns_

-   `import('lodash').DebouncedFunc<TFunc>`: Debounced function.

### useFocusableIframe

Dispatches a bubbling focus event when the iframe receives focus. Use
`onFocus` as usual on the iframe or a parent element.

_Returns_

-   `Object`: Ref to pass to the iframe.

### useFocusOnMount

Hook used to focus the first tabbable element on mount.

_Usage_

```js
import { useFocusOnMount } from '@wordpress/compose';

const WithFocusOnMount = () => {
	const ref = useFocusOnMount();
	return (
		<div ref={ ref }>
			<Button />
			<Button />
		</div>
	);
};
```

_Parameters_

-   _focusOnMount_ `boolean | 'firstElement'`: Focus on mount mode.

_Returns_

-   `import('react').RefCallback<HTMLElement>`: Ref callback.

### useFocusReturn

When opening modals/sidebars/dialogs, the focus
must move to the opened area and return to the
previously focused element when closed.
The current hook implements the returning behavior.

_Usage_

```js
import { useFocusReturn } from '@wordpress/compose';

const WithFocusReturn = () => {
	const ref = useFocusReturn();
	return (
		<div ref={ ref }>
			<Button />
			<Button />
		</div>
	);
};
```

_Parameters_

-   _onFocusReturn_ `[() => void]`: Overrides the default return behavior.

_Returns_

-   `import('react').RefCallback<HTMLElement>`: Element Ref.

### useInstanceId

Provides a unique instance ID.

_Parameters_

-   _object_ `object`: Object reference to create an id for.
-   _prefix_ `[string]`: Prefix for the unique id.
-   _preferredId_ `[string | number]`: Default ID to use.

_Returns_

-   `string | number`: The unique instance id.

### useIsomorphicLayoutEffect

Preferred over direct usage of `useLayoutEffect` when supporting
server rendered components (SSR) because currently React
throws a warning when using useLayoutEffect in that environment.

### useKeyboardShortcut

Attach a keyboard shortcut handler.

_Related_

-   <https://craig.is/killing/mice#api.bind> for information about the `callback` parameter.

_Parameters_

-   _shortcuts_ `string[]|string`: Keyboard Shortcuts.
-   _callback_ `(e: import('mousetrap').ExtendedKeyboardEvent, combo: string) => void`: Shortcut callback.
-   _options_ `WPKeyboardShortcutConfig`: Shortcut options.

### useMediaQuery

Runs a media query and returns its value when it changes.

_Parameters_

-   _query_ `[string]`: Media Query.

_Returns_

-   `boolean`: return value of the media query.

### useMergeRefs

Merges refs into one ref callback.

It also ensures that the merged ref callbacks are only called when they
change (as a result of a `useCallback` dependency update) OR when the ref
value changes, just as React does when passing a single ref callback to the
component.

As expected, if you pass a new function on every render, the ref callback
will be called after every render.

If you don't wish a ref callback to be called after every render, wrap it
with `useCallback( callback, dependencies )`. When a dependency changes, the
old ref callback will be called with `null` and the new ref callback will be
called with the same value.

To make ref callbacks easier to use, you can also pass the result of
`useRefEffect`, which makes cleanup easier by allowing you to return a
cleanup function instead of handling `null`.

It's also possible to _disable_ a ref (and its behaviour) by simply not
passing the ref.

```jsx
const ref = useRefEffect( ( node ) => {
  node.addEventListener( ... );
  return () => {
    node.removeEventListener( ... );
  };
}, [ ...dependencies ] );
const otherRef = useRef();
const mergedRefs useMergeRefs( [
  enabled && ref,
  otherRef,
] );
return <div ref={ mergedRefs } />;
```

_Parameters_

-   _refs_ `Array<TRef>`: The refs to be merged.

_Returns_

-   `import('react').RefCallback<TypeFromRef<TRef>>`: The merged ref callback.

### usePrevious

Use something's value from the previous render.
Based on <https://usehooks.com/usePrevious/>.

_Parameters_

-   _value_ `T`: The value to track.

_Returns_

-   `T | undefined`: The value from the previous render.

### useReducedMotion

Hook returning whether the user has a preference for reduced motion.

_Returns_

-   `boolean`: Reduced motion preference value.

### useRefEffect

Effect-like ref callback. Just like with `useEffect`, this allows you to
return a cleanup function to be run if the ref changes or one of the
dependencies changes. The ref is provided as an argument to the callback
functions. The main difference between this and `useEffect` is that
the `useEffect` callback is not called when the ref changes, but this is.
Pass the returned ref callback as the component's ref and merge multiple refs
with `useMergeRefs`.

It's worth noting that if the dependencies array is empty, there's not
strictly a need to clean up event handlers for example, because the node is
to be removed. It _is_ necessary if you add dependencies because the ref
callback will be called multiple times for the same node.

_Parameters_

-   _callback_ `( node: TElement ) => ( () => void ) | undefined`: Callback with ref as argument.
-   _dependencies_ `DependencyList`: Dependencies of the callback.

_Returns_

-   `RefCallback< TElement | null >`: Ref callback.

### useResizeObserver

Hook which allows to listen the resize event of any target element when it changes sizes.
_Note: `useResizeObserver` will report `null` until after first render_

Simply a re-export of `react-resize-aware` so refer to its documentation <https://github.com/FezVrasta/react-resize-aware>
for more details.

_Related_

-   <https://github.com/FezVrasta/react-resize-aware>

_Usage_

```js
const App = () => {
	const [ resizeListener, sizes ] = useResizeObserver();

	return (
		<div>
			{ resizeListener }
			Your content here
		</div>
	);
};
```

### useThrottle

Throttles a function with Lodash's `throttle`. A new throttled function will
be returned and any scheduled calls cancelled if any of the arguments change,
including the function to throttle, so please wrap functions created on
render in components in `useCallback`.

_Related_

-   <https://docs-lodash.com/v4/throttle/>

_Parameters_

-   _fn_ `TFunc`: The function to throttle.
-   _wait_ `[number]`: The number of milliseconds to throttle invocations to.
-   _options_ `[import('lodash').ThrottleSettings]`: The options object. See linked documentation for details.

_Returns_

-   `import('lodash').DebouncedFunc<TFunc>`: Throttled function.

### useViewportMatch

Returns true if the viewport matches the given query, or false otherwise.

_Usage_

```js
useViewportMatch( 'huge', '<' );
useViewportMatch( 'medium' );
```

_Parameters_

-   _breakpoint_ `WPBreakpoint`: Breakpoint size name.
-   _operator_ `[WPViewportOperator]`: Viewport operator.

_Returns_

-   `boolean`: Whether viewport matches query.

### useWarnOnChange

Hook that performs a shallow comparison between the preview value of an object
and the new one, if there's a difference, it prints it to the console.
this is useful in performance related work, to check why a component re-renders.

_Usage_

```jsx
function MyComponent( props ) {
	useWarnOnChange( props );

	return 'Something';
}
```

_Parameters_

-   _object_ `object`: Object which changes to compare.
-   _prefix_ `string`: Just a prefix to show when console logging.

### withGlobalEvents

> **Deprecated**

Higher-order component creator which, given an object of DOM event types and
values corresponding to a callback function name on the component, will
create or update a window event handler to invoke the callback when an event
occurs. On behalf of the consuming developer, the higher-order component
manages unbinding when the component unmounts, and binding at most a single
event handler for the entire application.

_Parameters_

-   _eventTypesToHandlers_ `Record<keyof GlobalEventHandlersEventMap, string>`: Object with keys of DOM event type, the value a name of the function on the original component's instance which handles the event.

_Returns_

-   `any`: Higher-order component.

### withInstanceId

A Higher Order Component used to be provide a unique instance ID by
component.

_Type_

-   `PropInjectingHigherOrderComponent< { instanceId: string | number; } >`

### withSafeTimeout

A higher-order component used to provide and manage delayed function calls
that ought to be bound to a component's lifecycle.

_Type_

-   `PropInjectingHigherOrderComponent< TimeoutProps >`

### withState

> **Deprecated** Use `useState` instead.

A Higher Order Component used to provide and manage internal component state
via props.

_Parameters_

-   _initialState_ `any`: Optional initial state of the component.

_Returns_

-   `any`: A higher order component wrapper accepting a component that takes the state props + its own props + `setState` and returning a component that only accepts the own props.

<!-- END TOKEN(Autogenerated API docs) -->

<br/><br/><p align="center"><img src="https://s.w.org/style/images/codeispoetry.png?1" alt="Code is Poetry." /></p>
