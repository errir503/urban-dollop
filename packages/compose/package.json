{
	"name": "@wordpress/compose",
	"version": "6.15.0",
	"description": "WordPress higher-order components (HOCs).",
	"author": "The WordPress Contributors",
	"license": "GPL-2.0-or-later",
	"keywords": [
		"wordpress",
		"gutenberg",
		"react",
		"hoc",
		"hooks"
	],
	"homepage": "https://github.com/WordPress/gutenberg/tree/HEAD/packages/compose/README.md",
	"repository": {
		"type": "git",
		"url": "https://github.com/WordPress/gutenberg.git",
		"directory": "packages/compose"
	},
	"bugs": {
		"url": "https://github.com/WordPress/gutenberg/issues"
	},
	"engines": {
		"node": ">=12"
	},
	"main": "build/index.js",
	"module": "build-module/index.js",
	"react-native": "src/index",
	"types": "build-types",
	"sideEffects": false,
	"dependencies": {
		"@babel/runtime": "^7.16.0",
		"@types/mousetrap": "^1.6.8",
		"@wordpress/deprecated": "file:../deprecated",
		"@wordpress/dom": "file:../dom",
		"@wordpress/element": "file:../element",
		"@wordpress/is-shallow-equal": "file:../is-shallow-equal",
		"@wordpress/keycodes": "file:../keycodes",
		"@wordpress/priority-queue": "file:../priority-queue",
		"change-case": "^4.1.2",
		"clipboard": "^2.0.8",
		"mousetrap": "^1.6.5",
		"use-memo-one": "^1.1.1"
	},
	"peerDependencies": {
		"react": "^18.0.0"
	},
	"publishConfig": {
		"access": "public"
	}
}
