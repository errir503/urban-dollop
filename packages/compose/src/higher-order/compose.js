/**
 * External dependencies
 */
import { flowRight as compose } from 'lodash';

/**
 * Composes multiple higher-order components into a single higher-order component. Performs right-to-left function
 * composition, where each successive invocation is supplied the return value of the previous.
 *
 * This is just a re-export of `lodash`'s `flowRight` function.
 *
 * @see https://docs-lodash.com/v4/flow-right/
 */
export default compose;
