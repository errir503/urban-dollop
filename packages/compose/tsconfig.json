{
	"extends": "../../tsconfig.base.json",
	"compilerOptions": {
		"rootDir": "src",
		"declarationDir": "build-types"
	},
	"include": [
		"src/higher-order/if-condition/**/*",
		"src/higher-order/pure/**/*",
		"src/utils/**/*"
	]
}
