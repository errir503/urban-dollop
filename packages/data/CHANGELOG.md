<!-- Learn how to maintain this file at https://github.com/WordPress/gutenberg/tree/HEAD/packages#maintaining-changelogs. -->

## Unreleased

## 6.8.0 (2022-05-04)

## 6.7.0 (2022-04-21)

## 6.6.0 (2022-04-08)

## 6.5.0 (2022-03-23)

## 6.4.0 (2022-03-11)

## 6.3.0 (2022-02-23)

### New Features

- Enabled thunks by default for all stores and removed the `__experimentalUseThunks` flag.
- Store the resolution errors in store metadata and expose them using `hasResolutionFailed` the `getResolutionError` meta-selectors ([#38669](https://github.com/WordPress/gutenberg/pull/38669)).
- Expose the resolution status (undefined, resolving, finished, error) via the `getResolutionState` meta-selector ([#38669](https://github.com/WordPress/gutenberg/pull/38669)).

## 6.2.1 (2022-02-10)

### Bug Fix

-   Removed unused `memize` dependency ([#38388](https://github.com/WordPress/gutenberg/pull/38388)).

## 6.2.0 (2022-01-27)

### Bug Fix

-   Corrected expect type of action creators and selectors in Redux store configuration type
-   Move `redux` to regular dependencies and update it to version `^4.1.2`.

### Internal

-   Changed names of store-related types to better reflect their use and role.
-   Changed "storeDefinition" to "storeDescriptor" to better reflect its use and role.

## 6.1.0 (2021-09-09)

### New Features

-   Added a `batch` registry method to batch dispatch calls for performance reasons.
-   Add a new migration for the persistence plugin to migrate edit-widgets preferences to the interface package. As part of this change deprecated migrations for the persistence plugin have been removed ([#33774](https://github.com/WordPress/gutenberg/pull/33774)).
-   Update data controls to accept a data store definition as their first param in addition to a string-based store name value ([#34170](https://github.com/WordPress/gutenberg/pull/34170)).

## 6.0.0 (2021-07-29)

### Breaking Change

-   Upgraded React components to work with v17.0 ([#29118](https://github.com/WordPress/gutenberg/pull/29118)). There are no new features in React v17.0 as explained in the [blog post](https://reactjs.org/blog/2020/10/20/react-v17.html).

## 5.2.0 (2021-07-21)

## 5.1.0 (2021-05-20)

## 5.0.0 (2021-05-14)

### Breaking Changes

-   Drop support for Internet Explorer 11 ([#31110](https://github.com/WordPress/gutenberg/pull/31110)). Learn more at https://make.wordpress.org/core/2021/04/22/ie-11-support-phase-out-plan/.
-   Increase the minimum Node.js version to v12 matching Long Term Support releases ([#31270](https://github.com/WordPress/gutenberg/pull/31270)). Learn more at https://nodejs.org/en/about/releases/.

### New Features

-   Added new `startResolutions` and `finishResolutions` actions as batched variants of `startResolution` and `finishResolutions` actions.

### Enhancements

-   Updated `redux` dependency from requiring `^4.0.0` to `^4.1.0` (see changes at https://github.com/reduxjs/redux/releases/tag/v4.1.0).

## 4.27.0 (2021-03-17)

### New Features

-   Added new `resolveSelect` registry method to initiate and wait for selector resolution

## 4.26.0 (2020-12-17)

### New Features

-   Added new `register` function for registering a standard `@wordpress/data` store definition ([#26655](https://github.com/WordPress/gutenberg/pull/26655)).
-   Added new `createReduxStore` factory function that creates a data store definition for the provided Redux store options to use with `register` function ([#26655](https://github.com/WordPress/gutenberg/pull/26655)).
-   Extended `select` and `dispatch` functions to accept a data store definition as their first param in addition to a string-based store name value [#26655](https://github.com/WordPress/gutenberg/pull/26655)).
-   Extended `useDispatch` hook to accept a data store definition as their first param in addition to a string-based store name value [#26655](https://github.com/WordPress/gutenberg/pull/26655)).

### Deprecations

-   `registerGenericStore` has been deprecated. Use `register` instead.
-   `registerStore` has been deprecated. Use `register` instead.

## 4.6.0 (2019-06-12)

### New Feature

-   Expose `useSelect` hook for usage in functional components. ([#15737](https://github.com/WordPress/gutenberg/pull/15737))
-   Expose `useDispatch` hook for usage in functional components. ([#15896](https://github.com/WordPress/gutenberg/pull/15896))

### Enhancements

-   `withSelect` internally uses the new `useSelect` hook. ([#15737](https://github.com/WordPress/gutenberg/pull/15737). **Note:** This _could_ impact performance of code using `withSelect` in edge-cases. To avoid impact, memoize passed in `mapSelectToProps` callbacks or implement `useSelect` directly with dependencies.
-   `withDispatch` internally uses a new `useDispatchWithMap` hook (an internal only api) ([#15896](https://github.com/WordPress/gutenberg/pull/15896))

## 4.5.0 (2019-05-21)

### Bug Fix

-   Restore functionality of action-generators returning a Promise. Clarify intent and behaviour for `dispatch` behaviour. Dispatch actions now always
    return a promise ([#14830](https://github.com/WordPress/gutenberg/pull/14830)

### Enhancements

-   Expose `hasResolver` property on returned selectors indicating whether the selector has a corresponding resolver.

## 4.3.0 (2019-03-06)

### Enhancements

-   The `registerStore` function now accepts an optional `initialState` option value.
-   Introduce new `invalidateResolutionForStore` dispatch action for signalling to invalidate the resolution cache for an entire given store.
-   Introduce new `invalidateResolutionForStoreSelector` dispatch action for signalling to invalidate the resolution cache for a store selector (and all variations of arguments on that selector).

### Bug Fix

-   Resolves issue in the persistence plugin where passing `persist` as an array of reducer keys would wrongly replace state values for the unpersisted reducer keys.
-   Restores a behavior in the persistence plugin where a default state provided as an object will be deeply merged as a base for the persisted value. This allows for a developer to include additional new keys in a persisted value default in future iterations of their store.

## 4.2.0 (2019-01-03)

### Enhancements

-   Optimized performance of selector execution (~511% improvement)

## 4.1.0 (2018-12-12)

### New Feature

-   `withDispatch`'s `mapDispatchToProps` function takes the `registry` object as the 3rd param ([#11851](https://github.com/WordPress/gutenberg/pull/11851)).
-   `withSelect`'s `mapSelectToProps` function takes the `registry` object as the 3rd param ([#11851](https://github.com/WordPress/gutenberg/pull/11851)).

## 4.0.1 (2018-11-20)

## 4.0.0 (2018-11-15)

### Breaking Changes

-   `registry.registerReducer` has been removed. Use `registry.registerStore` instead.
-   `registry.registerSelectors` has been removed. Use `registry.registerStore` instead.
-   `registry.registerActions` has been removed. Use `registry.registerStore` instead.
-   `registry.registerResolvers` has been removed. Use `registry.registerStore` instead.

### Bug Fix

-   Resolve an issue where `withSelect`'s `mapSelectToProps` would not be rerun if the wrapped component had incurred a store change during its mount lifecycle.

## 3.1.2 (2018-11-09)

## 3.1.1 (2018-11-09)

## 3.1.0 (2018-11-03)

### New Features

-   `registry.registerGenericStore` has been added to support integration with existing data systems.

### Deprecations

-   `registry.registerReducer` has been deprecated. Use `registry.registerStore` instead.
-   `registry.registerSelectors` has been deprecated. Use `registry.registerStore` instead.
-   `registry.registerActions` has been deprecated. Use `registry.registerStore` instead.
-   `registry.registerResolvers` has been deprecated. Use `registry.registerStore` instead.

## 3.0.1 (2018-10-30)

### Internal

-   Replace Redux implementation of `combineReducers` with in-place-compatible `turbo-combine-reducers`.

## 3.0.0 (2018-10-29)

### Breaking Changes

-   Writing resolvers as async generators has been removed. Use the controls plugin instead.

## 2.1.4 (2018-10-19)

## 2.1.3 (2018-10-18)

## 2.1.0 (2018-09-30)

### New Features

-   Adding support for using controls in resolvers using the controls plugin.

### Polish

-   Updated `redux` dependency to the latest version.

### Deprecations

-   Writing resolvers as async generators has been deprecated. Use the controls plugin instead.

### Bug Fixes

-   Fix the promise middleware in Firefox.

## 2.0.0 (2018-09-05)

### Breaking Change

-   The `withRehdyration` function is removed. Use the persistence plugin instead.
-   The `loadAndPersist` function is removed. Use the persistence plugin instead.
-   `restrictPersistence`, `setPersistenceStorage` and `setupPersistence` functions have been removed. Please use the data persistence plugin instead.
-   Change how required built-ins are polyfilled with Babel 7 ([#9171](https://github.com/WordPress/gutenberg/pull/9171)). If you're using an environment that has limited or no support for ES2015+ such as lower versions of IE then using [core-js](https://github.com/zloirock/core-js) or [@babel/polyfill](https://babeljs.io/docs/en/next/babel-polyfill) will add support for these methods.
