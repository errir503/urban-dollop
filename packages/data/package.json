{
	"name": "@wordpress/data",
	"version": "5.1.0",
	"description": "Data module for WordPress.",
	"author": "The WordPress Contributors",
	"license": "GPL-2.0-or-later",
	"keywords": [
		"wordpress",
		"gutenberg",
		"data",
		"redux"
	],
	"homepage": "https://github.com/WordPress/gutenberg/tree/HEAD/packages/data/README.md",
	"repository": {
		"type": "git",
		"url": "https://github.com/WordPress/gutenberg.git",
		"directory": "packages/data"
	},
	"bugs": {
		"url": "https://github.com/WordPress/gutenberg/issues"
	},
	"engines": {
		"node": ">=12"
	},
	"main": "build/index.js",
	"module": "build-module/index.js",
	"react-native": "src/index",
	"sideEffects": false,
	"dependencies": {
		"@babel/runtime": "^7.13.10",
		"@wordpress/compose": "file:../compose",
		"@wordpress/deprecated": "file:../deprecated",
		"@wordpress/element": "file:../element",
		"@wordpress/is-shallow-equal": "file:../is-shallow-equal",
		"@wordpress/priority-queue": "file:../priority-queue",
		"@wordpress/redux-routine": "file:../redux-routine",
		"equivalent-key-map": "^0.2.2",
		"is-promise": "^4.0.0",
		"lodash": "^4.17.21",
		"memize": "^1.1.0",
		"redux": "^4.1.0",
		"turbo-combine-reducers": "^1.0.2",
		"use-memo-one": "^1.1.1"
	},
	"publishConfig": {
		"access": "public"
	}
}
