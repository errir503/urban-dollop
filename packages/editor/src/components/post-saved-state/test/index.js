/**
 * External dependencies
 */
import { mount, shallow } from 'enzyme';

/**
 * WordPress dependencies
 */
import { useViewportMatch } from '@wordpress/compose';
import { useSelect } from '@wordpress/data';

/**
 * Internal dependencies
 */
import PostSavedState from '../';

const mockSavePost = jest.fn();

jest.mock( '@wordpress/data/src/components/use-dispatch', () => {
	return {
		useDispatch: () => ( { savePost: mockSavePost } ),
	};
} );

jest.mock( '@wordpress/data/src/components/use-select', () => {
	// This allows us to tweak the returned value on each test
	const mock = jest.fn();
	return mock;
} );

jest.mock( '@wordpress/compose/src/hooks/use-viewport-match', () => {
	// This allows us to tweak the returned value on each test
	const mock = jest.fn();
	return mock;
} );

describe( 'PostSavedState', () => {
	it( 'should display saving while save in progress, even if not saveable', () => {
		useSelect.mockImplementation( () => ( {
			isDirty: false,
			isNew: true,
			isSaveable: false,
			isSaving: true,
		} ) );

		const wrapper = mount( <PostSavedState /> );

		expect( wrapper.text() ).toContain( 'Saving' );
	} );

	it( 'returns null if the post is not saveable', () => {
		useSelect.mockImplementation( () => ( {
			isDirty: false,
			isNew: true,
			isSaveable: false,
			isSaving: false,
		} ) );

		const wrapper = shallow( <PostSavedState /> );

		expect( wrapper.type() ).toBeNull();
	} );

	it( 'returns a switch to draft link if the post is published', () => {
		useSelect.mockImplementation( () => ( {
			isPublished: true,
		} ) );

		const wrapper = shallow( <PostSavedState /> );

		expect( wrapper ).toMatchSnapshot();
	} );

	it( 'should return Saved text if not new and not dirty', () => {
		useSelect.mockImplementation( () => ( {
			isDirty: false,
			isNew: false,
			isSaveable: true,
			isSaving: false,
		} ) );

		const wrapper = shallow( <PostSavedState /> );

		expect( wrapper.childAt( 0 ).name() ).toBe( 'Icon' );
		expect( wrapper.childAt( 1 ).text() ).toBe( 'Saved' );
	} );

	it( 'should return Save button if edits to be saved', () => {
		useSelect.mockImplementation( () => ( {
			isDirty: true,
			isNew: false,
			isSaveable: true,
			isSaving: false,
		} ) );

		// Simulate the viewport being considered large.
		useViewportMatch.mockImplementation( () => true );

		const wrapper = shallow( <PostSavedState /> );

		expect( wrapper ).toMatchSnapshot();
		wrapper.simulate( 'click', {} );
		expect( mockSavePost ).toHaveBeenCalled();
		// Regression: Verify the event object is not passed to prop callback.
		expect( mockSavePost.mock.calls[ 0 ] ).toEqual( [] );
	} );
} );
