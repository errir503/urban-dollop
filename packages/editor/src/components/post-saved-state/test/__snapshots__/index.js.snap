// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostSavedState returns a switch to draft link if the post is published 1`] = `<WithSelect(WithDispatch(PostSwitchToDraftButton)) />`;

exports[`PostSavedState should return Save button if edits to be saved 1`] = `
<ForwardRef(Button)
  className="editor-post-save-draft"
  isTertiary={true}
  onClick={[Function]}
  shortcut="Ctrl+S"
>
  Save draft
</ForwardRef(Button)>
`;
